### DISCUSSION TITLE
Keyboard shortcuts for navigating tabs in preferences

### DISCUSSION THREAD

**Discussion by rak-phillip** [rancher-sandbox/rancher-desktop#3641] [Research]
We settled on primary navigation for the Preferences window in #2627, but still need to agree on an acceptable pattern for shortcuts that navigate tabs inside of preferences

![image](https://user-images.githubusercontent.com/835961/207971299-22ee66c6-0022-46b3-a8ea-b107e4781312.png)

We had proposed two options

1. Holding down a modifier key to initiate a keyboard shortcut will highlight shortcuts on the screen (e.g. underlining the associated letter in a menu item in a window menu)
2. ctrl/cmd + Tab to cycle tabs on a page

I'd like to discuss more to see if we can't settle on an optimal pattern

**Comment by jandubois** [comment]
> 1. Holding down a modifier key to initiate a keyboard shortcut will highlight shortcuts on the screen (e.g. underlining the associated letter in a menu item in a window menu)

I think this was in general a mechanism to help shortcut discoverability.

I thought we would be using number shortcuts for secondary (tab) selection, just like we do for primary.

My suggestion:

* For macOS use `⌘⌥1` (`Cmd + Option + 1`) etc. There is precedent for this in the Safari browser, which uses these keys to navigate to entries in the bookmark bar.

* For Windows (and Linux) use `Ctrl + Alt + 1` etc.  I couldn't find any precedent, but it feels like a logical extension. In many Microsoft Office apps (and Google Docs) these shortcuts are used for heading style 1 through 6.

FWIW, I used https://defkey.com/search?irq=Ctrl+%2B+alt+%2B+1 to explore what other apps are doing with this shortcut.

**Reply by mook-as** [reply]
From that same site, `Ctrl + 1` through `8` are generally used in browsers to jump to specific tabs. `Ctrl + 9` jumps to the last tab, `Ctrl + 0` resets zoom, and `Ctrl + -` / `Ctrl + =` zooms out and in, respectively.

Note that `⌘1` seems to jump to tab as well, including in Safari; ⌥ appears to be unnecessary. The other numbers also match up to the above.

**Reply by jandubois** [reply]
Yes, that is the "primary" navigation; we are using it to switch between "pages". So in the app `⌘1` will be the "General" page, `⌘2` the "Port Forwarding" page etc.

In the Preferences dialog `⌘1` will be the "Application" page, `⌘2` the "Virtual Machine" page etc.

But once you are on the "Application" page in Preferences, we want to have shortcuts to move from the "Behavior" tab to the "Environment" tab using "secondary" navigation shortcuts.

So to get from the app to the "Environment" tab you would press `⌘,` then `⌘1`, and finally `⌘⌥2` to bring up Preferences, switch to Application page, and move to Environment tab respectively.

And yes, `Ctrl + 1` etc. are the corresponding primary shortcuts on Windows and Linux. We already have agreement on using them for pages, both at the app level, and inside the Preferences dialog.

This discussion is about finding the best choices for the secondary navigation shortcuts to switch between tabs on a page.

<hr>

Some may find it confusing that we use the primary navigation keys for pages (vertical list) instead of tabs, like in a browser (horizontal list), but the criteria is really about navigating between the primary functionality areas in an app.

For example in Mail.app `⌘1` goes to your Inbox, `⌘2` to the Sent folder, `⌘3` to the Flagged folder etc.

**Comment by jandubois** [comment]
I guess for macOS I could also see `⌃1` working for secondary navigation, but it can be confusing because it would be the primary navigation on Windows and Linux.

`⌥1` etc would not work on macOS because those are regular character that you may want to enter into a text entry field. E.g. `⌥3` on a US keyboard would be the `£` character.

If we want to go with `⌃1` etc on macOS, then I think `Alt + 1` etc on Windows/Linux would be a possible choice as well.

I don't really have good arguments for picking one set over the other...

**Comment by jandubois** [comment]
While we are not looking (yet) for ways to highlight shortcuts to individual input fields, we would like to have some way to discover the secondary navigation shortcuts from within the app.

The primary shortcuts are based on platform conventions, so users can discover them on their own by simply trying the familiar combinations, but for secondary shortcuts we are just making them up.

We can put them in the docs (and once in release notes), but many users will miss them.

I guess we eventually want to have a docs page "Keyboard shortcuts" and a corresponding entry in the Help menu.
